contracts/token/ERC1155/ERC1155.sol

```diff
@@ -143,44 +143,75 @@ contract ERC1155 is Context, ERC165, IERC1155, IERC1155MetadataURI {
+        uint256[] memory ids,
+        uint256[] memory amounts,
         bytes memory data
     ) internal virtual {
-        require(to != address(0), "ERC1155: transfer to the zero address");
+        require(ids.length == amounts.length, "ERC1155: ids and amounts length mismatch");
 
         address operator = _msgSender();
-        uint256[] memory ids = _asSingletonArray(id);
-        uint256[] memory amounts = _asSingletonArray(amount);
 
-        _beforeTokenTransfer(operator, from, to, ids, amounts, data);
-
-        uint256 fromBalance = _balances[id][from];
-        require(fromBalance >= amount, "ERC1155: insufficient balance for transfer");
-        unchecked {
-            _balances[id][from] = fromBalance - amount;
-        }
-        _balances[id][to] += amount;
+        for (uint256 i = 0; i < ids.length; ++i) {
+            uint256 id = ids[i];
+            uint256 amount = amounts[i];
 
-        emit TransferSingle(operator, from, to, id, amount);
+            if (from != address(0)) {
+                uint256 fromBalance = _balances[id][from];
+                require(fromBalance >= amount, "ERC1155: insufficient balance for transfer");
+                unchecked {
+                    _balances[id][from] = fromBalance - amount;
+                }
+            }
 
-        _afterTokenTransfer(operator, from, to, ids, amounts, data);
+            if (to != address(0)) {
+                _balances[id][to] += amount;
+            }
+        }
+        if (ids.length == 1) {
+            uint256 id = ids[0];
+            uint256 amount = amounts[0];
+            emit TransferSingle(operator, from, to, id, amount);
+            if (to != address(0)) {
+                _doSafeTransferAcceptanceCheck(operator, from, to, id, amount, data);
+            }
+        } else {
+            emit TransferBatch(operator, from, to, ids, amounts);
+            if (to != address(0)) {
+                _doSafeBatchTransferAcceptanceCheck(operator, from, to, ids, amounts, data);
+            }
+        }
+    }
 
-        _doSafeTransferAcceptanceCheck(operator, from, to, id, amount, data);
+    /**
+     * @dev Transfers `amount` tokens of token type `id` from `from` to `to`.
+     *
+     * Emits a {TransferSingle} event.
+     *
+     * Requirements:
+     *
+     * - `to` cannot be the zero address.
+     * - `from` must have a balance of tokens of type `id` of at least `amount`.
+     * - If `to` refers to a smart contract, it must implement {IERC1155Receiver-onERC1155Received} and return the
+     * acceptance magic value.
+     */
+    function _safeTransferFrom(address from, address to, uint256 id, uint256 amount, bytes memory data) internal {
+        require(to != address(0), "ERC1155: transfer to the zero address");
+        require(from != address(0), "ERC1155: transfer from the zero address");
+        uint256[] memory ids = _asSingletonArray(id);
+        uint256[] memory amounts = _asSingletonArray(amount);
```

Comment:
We can probably gain some gas efficiency by defining a single function `_asSingletonArrays(id, amount) returns (uint256[] ids, uint256[] amounts)` that allocates the two arrays in assembly at the same time. We should use this function throughout the contract.